### ISSUE TITLE
Adding new parameter groups to an optimizer

### ISSUE THREAD

**Issue by Tetratrio** [NVIDIA/apex#292]
Hey,

I am currently training a growing model that requires me to add new param groups to the optimizer at each growth. That is, I train the model for X iterations, then add a new layer. This is repeated around 5-10 times. Preferably I keep the state of the optimizer for the older param groups when adding a new one.

Now I want to run this with mixed precision.
When using amp, there seem to be two options:
1. Add param group to optimizer where amp will take care of the operation. However, from my understanding, amp does not dynamically handle new parameter groups added to the optimizer?
2. Disable amp and restore the "pre-amp" state, add the param group and then reinitialize amp. I have only found a way to disable casts. If I try to reinitialize without first disabling amp, an error will be raised at https://github.com/NVIDIA/apex/blob/39e153a3159724432257a8fc118807b359f4d1c8/apex/amp/_process_optimizer.py#L257

I have not had time to go through all the code of this repository, but if I do I should probably be able to figure out how to manually reset amp. Although that would easily break with new apex releases and is not very elegant. So I thought I would check here first if there is an easy solution to my issue.

Thanks in advance!

P.S. I will be traveling for 2 days so I may not be able to respond to this issue.

**Comment by mcarilli** [comment]
Are you adding layers by unfreezing them (ie by changing their requires_grad attributes from False to True, so that they begin to receive gradients) or via an `optimizer.add_param_group` call?  

Amp should properly treat unfreezing or unfreezing (switching from requires_grad=True to False or the other way around) for any parameters as long as those parameters were seen by the optimizer's constructor (i.e., before the call to amp.initialize).

Adding param groups via `add_param_groups` after the call to amp.initialize is not currently supported but I could add support for it.

Finally, when you say you are "adding new layers" does that mean you are altering the model itself by adding layers to it on the fly, or is the model fully constructed from the beginning, and you are simply allowing more of the model's parameters to be seen/updated by the optimizer (by sending these params to the optimizer via `add_param_group` or by changing their requires_grad attributes to True)?

**Comment by Tetratrio** [comment]
I am adding layers on the fly, but I should be able to construct the entire network if I make some changes in my code. This is the only time I have ever had the need for `add_param_group` and I can work around it anyway.

It is a feature I can live without but if you feel like adding it I would still use it. Maybe there are other types of more dynamically growing networks that need it?

**Comment by mcarilli** [comment]
Adding layers on the fly is something that would be challenging to support for all opt_levels, since those layers might need to be casted.

Supporting `add_param_group` for optimizers would be easier.  I'll tag this to track it.

Initially constructing your model and optimizer with all the eventually-intended params visible, but some of them frozen (requires_grad=False), then sending them through amp.initialize once and gradually unfreezing them (setting requires_grad=True at the desired point during training), is another way to achieve the same effect.  As I said, this is something that Amp should currently support so it may be a viable workaround.

**Comment by Tetratrio** [comment]
Hey, thanks for looking into this.
Should it be possible with opt-level 01? Would I need to do something more than adding the new param groups in this case?
I looked at constructing the entire network at the start, but it would require more time than I thought to refactor the code and I have not had enough time to take care of it yet.
Maybe I could instead use `add_param_group` as another type of workaround once this feature is added to amp? Although I would be locked to certain opt-levels, that is okay for this specific issue.

**Comment by mcarilli** [comment]
Merged https://github.com/NVIDIA/apex/pull/310.  If you pull and reinstall from the latest master, `optimizer.add_param_group` should now "just work" for any opt_level at any point during training.

However, Amp still requires that the models themselves be fully constructed with all intended parameters before they are passed through `amp.initialize`.  I don't see an easy workaround for that off the top of my head.

Let me know if this is enough for your use case.

**Comment by Tetratrio** [comment]
Hey, thanks for the quick implementation!

So there are no workarounds for specific opt levels such as 01 or 02 (I just need one of them)? I dont think my use case is very common, although there are some growing networks such as https://arxiv.org/pdf/1806.09055.pdf .
I don't think it warrants the large amount of work required to implement this functionality in apex.

If it was somehow possible to disable amp (recast weights, remove hooks e.t.c.) that would solve the issue as well because then I could just disable and then re enable amp with the new layers added to the network. Or maybe if I could initialize amp multiple times (as long as I am using the same opt level). But I have not really looked too deeply at the code for amp so I am unsure of the workload for adding any of these two functionalities.

I will have to rewrite my code but in the future I can just keep in mind that I should have all layers instantiated before training starts. So not a big problem.

Thanks again for the help, greatly appreciate it!
Feel free to close this issue :)

**Comment by mcarilli** [comment]
I'll think about what it would take to make amp.initialize callable multiple times (with the same model or with new models).  It might be straightforward to implement for a restricted subset of model changes.

**Comment by Tetratrio** [comment]
I found an easy way to change my code to allow for use with apex so I am in no need of this feature. Although if you want to support dynamically growing networks it would be necessary. But the low amounts of these networks may not warrant the time it takes to implement for apex. Its all up to you!
